streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -792,7 +790,16 @@ public void punctuate(final ProcessorNode<?, ?, ?, ?> node,
             throw new IllegalStateException(String.format("%sCurrent node is not null", logPrefix));
         }
 
-        updateProcessorContext(node, time.milliseconds(), null);
+        // when punctuating, we need to preserve the timestamp (this can be either system time or event time)
+        // while other record context are set as dummy: null topic, -1 partition, -1 offset and empty header
+        final ProcessorRecordContext recordContext = new ProcessorRecordContext(
+            timestamp,
+            -1L,
+            -1,
+            null,
+            new RecordHeaders()
```

Comment:
Why not pass `null` ?

Reply:
I'm following the existing behavior here: if the record context is `null`, we also return a `return new RecordHeaders();`.